### ISSUE TITLE
Early data was sent by s_client without psk (because session expired)

### ISSUE THREAD

**Issue by Neo-ZK** [openssl/openssl#13395]
Openssl will send early data even if there is no psk extension.

Use s_client tools can get it:
```
openssl s_client -connect 127.0.0.1:443  -sess_in s.lo -early_data httpget.txt -tls1_3
```
`s.lo` is the session ticket of last tls conversation, when it expire, client would recognized it and not send psk extension, but client would still send early data.

I'm not sure RFC8446 forced that client MUST send early data with psk,  but I'm sure this will save network bandwidth. :-)

**Comment by richsalz** [comment]
Yes, the RFC says not to do that; see https://tools.ietf.org/html/rfc8446#section-4.2.10

On the other hand, the openssl apps are often used for debugging, so it makes sense that `s_client` can do non-conformant things.

I suggest that this can be closed as long as it's possible to do the right thing.

**Comment by davidben** [comment]
It's worth checking how `s_client` got into such a state. An application which simply configures the session via `SSL_set_session` and then uses OpenSSL's APIs around 0-RTT should not accidentally do something invalid if OpenSSL internally decides not to offer the session.

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
I think it would be beneficial to spend a little time investigating this.
